502 - manipulate data in production (#674)
bcgov/CONN-CCBC-portal: docs/process_to_manipulate_data.md

### Process to manipulate data in production

> 

Reviewer: ```suggestion

To manipulate data in production, the default workflow is to create a `sqitch` change, associated with `pgTap` tests that are run in our CI pipeline and tested in our `dev` and `test` environments as the application is deployed.

In rare cases, the team needs to manipulate data directly in the database, without going through our usual guardrails. Examples of this are mostly around manipulation of private data. This repository is open-source, therefore we should not commit any private information.

This should be considered a last resort. Updating data in production could lead to cascading failure if done without the proper guardrails, which our CI are providing.
```

Author: Nice addition thanks.

#### Important steps before you begin:

> - Get request/approval from Product Owner
> - Back up database before manipulating data
> - Pair program with another developer

Reviewer: ```suggestion
- Get an email request from Product Owner, including a date request for the changes
- Post a message in the team's channel to find an other developer who is available to pair with you to make the change.
- Respond to the email request, ensuring that the work will not be done by multiple devs concurrently
- Start a meeting, sharing your screen with the other developer the entire time
- Back up database before manipulating data (add link to manual backup instructions here)
```

### Process to manipulate data in production

To manipulate data in production, the default workflow is to create a `sqitch` change, associated with `pgTap` tests that are run in our CI pipeline and tested in our `dev` and `test` environments as the application is deployed.

In rare cases, the team needs to manipulate data directly in the database, without going through our usual guardrails. Examples of this are mostly around manipulation of private data. This repository is open-source, therefore we should not commit any private information.

This should be considered a last resort. Updating data in production could lead to cascading failure if done without the proper guardrails, which our CI are providing.
#### Important steps before you begin:

- Get request/approval from Product Owner
- Back up database before manipulating data
- Pair program with another developer

#### Manipulating the data

> 

Reviewer: do we want a command line to copy-paste to find the leader pod?
Something like: 
```
oc project ff61fb-dev
oc -n ff61fb-dev get pods -l postgres-operator.crunchydata.com/cluster=ccbc,postgres-operator.crunchydata.com/role=master
oc -n ff61fb-dev exec -it <db-leader-pod-name> -- psql -d ccbc
```
And a note:

> The `oc project` command and `-n` option in subsequent commands are redundant. They act as an additional guardrail to prevent accidentally running commands in the wrong namespace

Reviewer 2: I'm for a copy-paste to find the leader pod

Author: Alright I added a section with this

Don't start in `prod`, always start in the `dev` namespace to practice and mitigate the chance of a failure in prod.

> 

Reviewer: ```suggestion

0. Read through the steps below before starting.

```

1. Don't commit changes as superuser, change to the database user:

`set role ccbc;`

2. Always do the changes in a transaction:

`begin;`

> 

Reviewer: ```suggestion

At this point, you should see the following `psql` prompt: `ccbc=*>`. The `*` indicates that you are in a transaction, and the `>` is the normal user prompt (instead of `#` for a superuser)
```

It can be pretty stressful to do things without guardrails, lets' reassure the reader that they've run their steps properly

<br />

3. Insert/update the data

4. If there was an error roll back the transaction:

`rollback;`

5. If the command was successful select the data and verify the data

> 6. If everything looks good commit the transaction:

Reviewer: ```suggestion
6. If everything looks good to you and the other developer(s) on the call, commit the transaction:
```
